core/src/org/samtools/htsjdk/core/io/ByteSource.java

```diff
@@ -0,0 +1,107 @@
+package org.samtools.htsjdk.core.io;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Interface for classes that produce binary data (i.e. bytes) that can be decoded
+ * into other datatypes.
+ *
+ * Methods should throw exceptions if:
+ * - The underlying data source/stream/channel is closed
+ * - There are insufficient bytes available (after blocking) to read a single item of the datatype requested
+ *
+ * NOTES:
+ *   - Almost all the methods here could have default implementations that read via one of the readBytes() methods.
+ *   - Should think through how to deal with unsigned types
+ *   - Need to think how this would work with BlockCompressed data
+ */
+public interface ByteSource extends AutoCloseable {
+    /** Returns a description of the underlying source of the bytes, e.g. a path, stdin, a URL etc. */
+    String getDescription();
+
+    // Methods related to buffering/repositioning/skipping etc.
+    /** Returns true if the source has a known finite length, false otherwise. */
+    boolean hasLength();
+
+    /** Returns the length if {@link #hasLength()} is true, otherwise throws an exception. */
+    default long length() { throw new UnsupportedOperationException(); }
+
+    /** Returns the offset of the next byte that will be read by any operation. */
+    long position();
```

Comment:
Is this going to be a virtual file pointer in the case of a compressed file?

Reply:
I won't recommend that: a compressed file should also be able to seek/getPosition from the disk-representation (I had some struggles about it with the bgzip implementation in the previous htsjdk)

Reply:
I was thinking that yes, `BlockCompressedInputStream` would become a sub-class of this and the positions it deals in would be virtual file pointers.  @magicDGS why wouldn't you recommend this?

Reply:
I have two concerns about handling position in bgzip (and other compressed formats) with virtual file pointers with the same interface:

* Position in compressed/uncompressed data mean different things depending on the implementation, which is error prone. The API definition should behave exactly the same in every implementation, at least at the lower-level.
* In the case of uncompressed data, the returned value is the file-position; in bgzip, it will be a virtual file pointer (block number and position within block); and what's about other compressed formats?


The concrete use case that I have in mind is the bgzip FASTA reference, which I implemented in the lates HTSJDK release. It requires a .gzi index file  to convert to the virtual file pointer the .fai coordinates. Nevertheless, an implementation to seek to a non-virtual file pointer is possible and it might support non-gzi-indexed files, and at the same time handle as a normal stream if interest to seek to a concrete file position is required. Separating both concepts is something benefitial in my opinion.

Reply:
Methods for seeking to a virtual file-pointer in bgzip and/or uncompressed-data pointer in other compressed format can be supported with a different method where the value passed is interpreted in an implementation-dependent way. That could be another interface to represent data that is transformed (compressed, block-compressed, maybe encrypted...)